### DISCUSSION TITLE
Under Sonoma (Mac M1) I get httpd error  256

### DISCUSSION THREAD

**Discussion by UweRasche** [Homebrew/discussions#4915] [Getting started]
### Output of `brew config`

```shell
uwerasche@raschesserver ~ % brew config
HOMEBREW_VERSION: 4.1.20
ORIGIN: https://github.com/Homebrew/brew
HEAD: d3d51d3448966b5250b5c60110c945040054803d
Last commit: 8 days ago
Core tap JSON: 17 Nov 20:57 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1.1-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.0.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: /usr/bin occurs before /opt/homebrew/bin in your PATH.
This means that system-provided programs will be used instead of those
provided by Homebrew. Consider setting your PATH so that
/opt/homebrew/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/opt/homebrew/bin:$PATH"' >> ~/.zshrc

The following tools exist at both paths:
  openssl
```

### Description of issue

Hello, 
I want to Install a SSL webserver under Sonoma at my Mac. I install Homebrew and httpd at my Mac. I hate the console.

I change the listen to 80 and the document path.

uwerasche@raschesserver ~ % brew services start httpd
==> Successfully started `httpd` (label: homebrew.mxcl.httpd)
uwerasche@raschesserver ~ % brew services list       
Name  Status  User      File
httpd started uwerasche ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist

No problem. In the httpd.conf I load only the module:

#Uwe:
LoadModule cache_socache_module lib/httpd/modules/mod_cache_socache.so

I get:
uwerasche@raschesserver ~ % brew services list       
Name  Status     User      File
httpd error  256 uwerasche ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist

In the next step:

uwerasche@raschesserver ~ % codesign --sign ${myName} --force --keychain ~/Library/Keychains/login.keychain-db /opt/homebrew/Cellar/httpd/2.4.58/lib/httpd/modules/mod_cache_socache.so

/opt/homebrew/Cellar/httpd/2.4.58/lib/httpd/modules/mod_cache_socache.so: replacing existing signature

uwerasche@raschesserver ~ % codesign -dv --verbose=4 "/opt/homebrew/Cellar/httpd/2.4.58/lib/httpd/modules/mod_cache_socache.so"
Executable=/opt/homebrew/Cellar/httpd/2.4.58/lib/httpd/modules/mod_cache_socache.so
Identifier=mod_cache_socache
Format=Mach-O thin (arm64)
CodeDirectory v=20400 size=757 flags=0x0(none) hashes=18+2 location=embedded
VersionPlatform=1
VersionMin=917504
VersionSDK=917504
Hash type=sha256 size=32
CandidateCDHash sha256=f0842d7ebafb5d389b207ce279659c219b107c22
CandidateCDHashFull sha256=f0842d7ebafb5d389b207ce279659c219b107c226d45dd47cf997ad818a3b3bb
Hash choices=sha256
CMSDigest=f0842d7ebafb5d389b207ce279659c219b107c226d45dd47cf997ad818a3b3bb
CMSDigestType=2
Executable Segment base=0
Executable Segment limit=32768
Executable Segment flags=0x0
Page size=4096
CDHash=f0842d7ebafb5d389b207ce279659c219b107c22
Signature size=8980
Authority=Developer ID Application: Rasche-Software (KH23P8U5C6)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=19. Nov 2023 at 10:18:15
Info.plist=not bound
TeamIdentifier=KH23P8U5C6
Sealed Resources=none
Internal requirements count=1 size=180

Now I get again the error 

uwerasche@raschesserver ~ % brew services start httpd
==> Successfully started `httpd` (label: homebrew.mxcl.httpd)
uwerasche@raschesserver ~ % brew services list       
Name  Status     User      File
httpd error  256 uwerasche ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist

What can I do ?

**Comment by SMillerDev** [comment]
Check the error log for httpd, it should show you what went wrong.

**Reply by UweRasche** [reply]
The file is empty. I`m surprised

**Reply by GeorgeCiesinski** [reply]
I have the same problem as UweRasche, down to the logs showing nothing. It turns out the apache logs are being redirected to my system logs and not to the homebrew apache logs. I have posted a reply below with much more information and all of my troubleshooting.

**Comment by GeorgeCiesinski** [comment]
I am having the same issue. I've actually been troubleshooting this for 2 weeks and I am frustrated as seemingly no solution can be found. 

# M1 Macbook Air

Sonoma 14.4.1

# Brew Config

```
HOMEBREW_VERSION: 4.3.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: e0bc557e7b991cb23583679e1cf1c8a92b793aeb
Last commit: 12 hours ago
Core tap JSON: 20 May 21:04 UTC
Core cask tap JSON: 20 May 21:04 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.3.1 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4.1-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: 15.4
Rosetta 2: false
```

# Brew Doctor

```
Your system is ready to brew.
```

# Startup

I ran `brew services start httpd` and got the message "==> Successfully started `httpd` (label: homebrew.mxcl.httpd)".

# Brew Services List

```
georgeciesinski@georges-air ~ % brew services list
Name  Status     User            File
httpd error  256 georgeciesinski ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist
```

# Troubleshooting

I have reinstalled brew httpd multiple times. Most recently, I uninstalled it and ran `rm -rf /opt/homebrew/etc/httpd` to get rid of all my config files that I changed. I reinstalled httpd and started it again, but no changes. 

# Error log

**1) httpd error_log:** This is weird because after the reinstall, my error log located in `/opt/homebrew/var/log/httpd/error_log` is no longer showing up to date logs. For example, I just reinstalled httpd using homebrew, and ran `brew services start httpd` at 17:18, yet the logs don't reflect this: 

```
tail -f /opt/homebrew/var/log/httpd/error_log

[Mon May 20 16:27:27.884002 2024] [mpm_prefork:notice] [pid 7651] AH00173: SIGHUP received.  Attempting to restart
[Mon May 20 16:27:27.909913 2024] [mpm_prefork:notice] [pid 7651] AH00163: Apache/2.4.59 (Unix) PHP/8.3.7 configured -- resuming normal operations
[Mon May 20 16:27:27.909929 2024] [core:notice] [pid 7651] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd'
[Mon May 20 16:53:01.210884 2024] [mpm_prefork:notice] [pid 7651] AH00169: caught SIGTERM, shutting down
```

**2) System Log Files:** I was however able to get more up to date logs by checking the system logs. `log show --predicate 'processImagePath contains "httpd"'` shows

```
2024-05-20 17:17:24.618569-0400 0xc2924    Error       0x0                  22467  0    httpd: (libsystem_info.dylib) [com.apple.network.libinfo:si_destination_compare] send failed: Invalid argument
2024-05-20 17:17:24.618573-0400 0xc2924    Error       0x0                  22467  0    httpd: (libsystem_info.dylib) [com.apple.network.libinfo:si_destination_compare] send failed: Invalid argument
2024-05-20 17:17:24.618576-0400 0xc2924    Error       0x0                  22467  0    httpd: (libsystem_info.dylib) [com.apple.network.libinfo:si_destination_compare] send failed: Invalid argument
2024-05-20 17:17:24.625558-0400 0xc2924    Activity    0x14e010             22467  0    httpd: (libsystem_info.dylib) Retrieve User by Name
2024-05-20 17:17:24.626095-0400 0xc2924    Activity    0x14e011             22467  0    httpd: (libsystem_info.dylib) Retrieve Group by Name
2024-05-20 17:17:24.627074-0400 0xc2924    Activity    0x14e012             22467  0    httpd: (libsystem_info.dylib) Retrieve User by ID
2024-05-20 17:17:24.634818-0400 0xc2924    Default     0x0                  22467  0    httpd: (Network) [com.apple.network:] networkd_settings_read_from_file initialized networkd settings by reading plist directly
2024-05-20 17:17:24.635155-0400 0xc2924    Default     0x0                  22467  0    httpd: (Network) [com.apple.network:] networkd_settings_read_from_file initialized networkd settings by reading plist directly
--------------------------------------------------------------------------------------------------------------------
Log      - Default:          7, Info:                0, Debug:             0, Error:         46, Fault:          0
Activity - Create:          94, Transition:          0, Actions:           0
```

# Summary

So after reinstalling httpd and checking the logs, it isn't clear what is going on. This works without issue on my M2 work macbook pro, but not on my personal M1 macbook air.

**Reply by GeorgeCiesinski** [reply]
I should also add that I am able to start httpd using `sudo apachectl start` but this shows "Status None" when I run `brew services list`. 

Stranger still, when I run `which apachectl` it shows me the homebrew version, and it uses my homebrew apache config. It even works fine and I can access my sites. Then when I run `sudo apachectl stop` and `brew services start httpd` it shows "Status Error 256" and now my sites don't work. 

I have no clue why this is happening.

**Comment by GeorgeCiesinski** [comment]
Sorry for the spam. I am updating because I believe I found a solution that may help someone in the future. This solution assumes that the **error_log does not show errors** or the errors are outdated. I have found many posts reporting the issue of logs being empty or showing outdated logs, and it is very frustrating because volunteers who try to help ask for logs and are unable to help when those logs are empty. 

For anyone else experiencing the problem of empty logs and Error 256, this is how I fixed the problem.

# Solution

1) When you run the below command,

```
brew info httpd
```

it outputs: 

> To restart httpd after an upgrade:
> brew services restart httpd
> Or, if you don't want/need a background service you can just run:
> /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND

In [another discussion](https://github.com/orgs/Homebrew/discussions/3525) I found the suggestion to try running this command and see the output. In my case: 

```
/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
```

> (13)Permission denied: AH00091: httpd: could not open error log file /opt/homebrew/var/log/httpd/error_log.

This indicated that httpd doesn't have permission to write to error_log. 

2) Next, cd into homebrew

```
cd /opt/homebrew/var/log/httpd/
```

Check permissions

```
ls -al
```

> -rw-r--r--  1 root             admin  29727 20 May 19:13 access_log
> -rw-r--r--  1 root             admin  22508 20 May 19:17 error_log

For anyone unclear about permissions, this shows that the `access_log` and `error_log` are owned by root, and that other users only have permission to read, not write to these files. I found some posts suggesting that the owner should be apache or httpd, but I wasn't sure which to use and kept looking. [This post](https://bobcares.com/blog/httpd-could-not-open-error-log-file/) suggests to change permissions to 666. 

```
sudo chmod 666 error_log
```

```
sudo chmod 666 access_log
```

This results in permissions -rw-rw-rw- which allows groups and users to write to the file, and by extension, whatever user httpd is using.

# Testing solution

Now when I run 

```
/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
``` 

there is no error in the terminal. 

In a second terminal, I ran 

```
tail -f /opt/homebrew/var/log/httpd/error_log
```
and it is showing the command I ran in the first terminal with the correct timestamp indicating it is working. I went back to the first terminal and killed process with `ctrl + c`, and started the service by running 

```
brew services start httpd
```

This finally caused the error to go away and be replaced with "started"

> httpd started georgeciesinski ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist

I restarted the computer and listed brew services again confirming it is still running. 

Finally, I navigated to localhost:8080 and see "It Works".
